specification/dartLangSpec.tex

```diff
@@ -673,26 +674,104 @@ \subsection{Scoping}
+That is, if \NamespaceName{} has the bindings
+$n\mapsto{}V_1$ and $n\mapsto{}V_2$
+then $V_1 = V_2$.%
+}
 
 \LMHash{}%
-A scope $S_0$ induces a namespace $NS_0$ that maps the simple name of each variable, type or function declaration $d$ declared in $S_0$ to $d$.
-Labels are not included in the induced namespace of a scope; instead they have their own dedicated namespace.
+Let \NamespaceName{} be a namespace.
+We say that a name $n$ \Index{is in} \NamespaceName{}
+if $n$ is a key of \NamespaceName{}.
+We say a declaration $d$ \NoIndex{is in} \NamespaceName{}
+if a key of \NamespaceName{} is mapped to $d$.
 
-\commentary{
-It is therefore impossible, e.g., to define a class that declares a method and a getter with the same name in Dart.
-Similarly one cannot declare a top-level function with the same name as a library variable or a class.
+\LMHash{}%
+A scope $S_0$ has an associated namespace \NamespaceName{0}.
+The bindings of \NamespaceName{0} is specified in this document by saying that
+a given declaration $D$ named $n$
+\IndexCustom{introduces}{declaration!introduces an entity into a scope}
+a specific entity $V$ into $S_0$,
+which means that the binding $n\mapsto{}V$ is added to \NamespaceName{0}.
+
+\commentary{%
+In some cases, the name of the declaration differs from
+the identifier that occurs in the declaration syntax used to declare it.
+Setters have names that are distinct from the corresponding getters
+because they always have an \lit{=} automatically added at the end,
+and the unary minus operator has the special name \code{unary-}.%
+}
+
+\commentary{%
+It is typically the case that $V$ is $D$,
+but there are exceptions:
+A variable declaration introduces an implicitly induced getter declaration,
+and in some cases also an implicitly induced setter declaration into the
+given scope.%
+}
+
+\commentary{%
+Note that labels are not included in the namespace of a scope.
+They are resolved lexically rather then being looked up in a namespace.%
 }
 
 \LMHash{}%
-It is a compile-time error if there is more than one entity with the same name declared in the same scope.
+It is a compile-time error if there is more than one entity with the same name
+declared in the same scope.
 
-\commentary{
-In some cases, the name of the declaration differs from the identifier used to declare it.
-Setters have names that are distinct from the corresponding getters because they always have an = automatically added at the end, and unary minus has the special name unary-.
+\commentary{%
+It is therefore impossible, e.g., to define a class that declares
+a method and a getter with the same name in Dart.
+Similarly one cannot declare a top-level function with
+the same name as a library variable or a class
+which is declared in the same library.%
+}
+
+\LMHash{}%
+We introduce the notion of a
+\Index{run-time namespace}.
+This is a partial function from names to run-time entities,
+in particular storage locations and compiled code.
```

Comment:
This is the usual conflict: I think it isn't helpful to imply that every use of every function in a Dart program must rely on a full-fledged function object. In reality, we can have a top-level function which is called all the time, but there is never a function object for it, because all those invocations are statically resolved.

But I changed it to `functions` which is weasel-wording for "something that's associated with a function declaration and which doesn't have to be a function object".